driver-core/src/main/com/mongodb/internal/connection/X509Authenticator.java

```diff
@@ -49,23 +53,35 @@ void authenticate(final InternalConnection connection, final ConnectionDescripti
     void authenticateAsync(final InternalConnection connection, final ConnectionDescription connectionDescription,
                            final SingleResultCallback<Void> callback) {
         try {
-            validateUserName(connectionDescription);
-            executeCommandAsync(getMongoCredential().getSource(), getAuthCommand(getMongoCredential().getUserName()), connection,
-                                new SingleResultCallback<BsonDocument>() {
-                                    @Override
-                                    public void onResult(final BsonDocument nonceResult, final Throwable t) {
-                                        if (t != null) {
-                                            callback.onResult(null, translateThrowable(t));
-                                        } else {
-                                            callback.onResult(null, null);
-                                        }
-                                    }
-                                });
+            if (this.speculativeAuthenticateResponse == null) {
+                validateUserName(connectionDescription);
+                executeCommandAsync(getMongoCredential().getSource(), getAuthCommand(getMongoCredential().getUserName()), connection,
+                        new SingleResultCallback<BsonDocument>() {
+                            @Override
+                            public void onResult(final BsonDocument nonceResult, final Throwable t) {
+                                if (t != null) {
+                                    callback.onResult(null, translateThrowable(t));
+                                } else {
+                                    callback.onResult(null, null);
+                                }
+                            }
+                        });
+            }
         } catch (Throwable t) {
             callback.onResult(null, t);
         }
     }
 
+    @Override
+    protected BsonDocument createSpeculativeAuthenticateCommand(final InternalConnection connection) {
+        return getAuthCommand(getMongoCredential().getUserName()).append("db", new BsonString("$external"));
```

Comment:
I think this method should call `validateUserName`.  Otherwise we don't validate it anymore